Allow multi-select NFTs

johnnyd-eth opened the pull request:
## What does this PR do and why?

https://user-images.githubusercontent.com/96150256/202636770-7ae3e02c-6e9b-494d-b23e-aba024fb1041.mp4

I've updated this one to default dontMint=true when no NFTs selected. I've tested and it's working well for me but feel free to give it a whirl yourself. 

This has removed the necessity for any linkage from the pay amount input to the NFT selection (still goes the other way) - I've done this.

Also snuck in a quick fix for the memo images (closes #2062)

## Acceptance checklist

- [ ] I have evaluated the [Approval Guidelines](https://github.com/jbx-protocol/juice-interface/blob/main/CONTRIBUTING.md#approval-guidelines) for this PR.
- [ ] I have tested this PR in [all supported browsers](https://github.com/jbx-protocol/juice-interface/blob/main/CONTRIBUTING.md#supported-browsers).
- [ ] I have tested this PR in dark mode and light mode (if applicable).

tomquirk commented on src/utils/nftRewards.ts:
```diff
@@ -292,6 +292,31 @@ export function encodeJB721DelegatePayMetadata(
+  rewardTiers: NftRewardTier[]
+  indices: number[]
+}): number[] {
+  return rewardTiers
+    .map((tier, idx) => {
+      if (indices.includes(idx)) {
+        return tier.id
+      }
+    })
+    .filter(Number) as number[]
```
**suggestion**: does flipping the logic do the same thing?

```suggestion
  return rewardTiers
    .filter((tier, idx) => indicies.includes(idx))
    .map(tier => tier.id)
```

johnnyd-eth replied:
Ah yes of course. Very nice

tomquirk commented on src/components/v2v3/V2V3Project/V2V3PayButton/V2V3ConfirmPayModal/V2V3ConfirmPayModal.tsx:
```diff
@@ -262,7 +261,7 @@ export function V2V3ConfirmPayModal({
                       You receive these NFTs for contributing{' '}
                       <strong>
-                        {nftRewardTier.contributionFloor} ETH or more
+                        {sumTierFloors(nftRewardTiers)} ETH or more
```
**suggestion** does it make sense to just remove `or more` now?

johnnyd-eth replied:
I think it still makes sense. You can still pay more than each tier individually and also more than the sum of any, e.g: 
<img width="584" alt="Screen Shot 2022-11-20 at 12 04 25 pm" src="https://user-images.githubusercontent.com/96150256/202879238-5bff4152-3c04-462a-81b1-d462d6657346.png">

tomquirk replied:
sgtm

tomquirk commented on src/components/v2v3/V2V3Project/V2V3PayButton/V2V3ConfirmPayModal/V2V3ConfirmPayModal.tsx:
```diff
@@ -95,11 +95,10 @@ export function V2V3ConfirmPayModal({
   })
 
-  let nftRewardTier: NftRewardTier | undefined
+  let nftRewardTiers: NftRewardTier[] | undefined
```
**suggestion** this can be combined into one line

```suggestion
  const nftRewardTiers: rewardTiers?.filter // etc.
```

johnnyd-eth replied:
Ah yep lol whack

tomquirk commented on src/components/v2v3/V2V3Project/V2V3PayButton/V2V3ConfirmPayModal/NftRewardCell.tsx:
```diff
@@ -1,24 +1,24 @@
   const [imageLoading, setImageLoading] = useState<boolean>(true)
 
-  const nftImage = (
+  const nftImage = (tier: NftRewardTier) => (
```
**suggestion** perhaps this can be its own component, declared outside of the scope of NftRewardCell (can probably still be in this file)

johnnyd-eth replied:
Yeah good call

tomquirk commented on src/components/NftRewards/NftRewardsSection.tsx:
```diff
@@ -172,9 +223,9 @@ export function NftRewardsSection() {
-                  isSelected={idx === selectedIndex}
-                  onClick={() => handleSelected(rewardTier, idx)}
-                  onRemove={deselectTier}
+                  isSelected={selectedIndices.includes(idx)}
```
**question (non-blocking)** I don't really understand why we dont use the tier ID instead of the index. Tier ID seems safer and less messy overall. WDYT?

johnnyd-eth replied:
Yeah I agree, I've overhauled the unnecessary indice logic

tomquirk commented on src/components/NftRewards/NftRewardsSection.tsx:
```diff
@@ -75,40 +78,69 @@ export function NftRewardsSection() {
   const payAmountETH =
     payInCurrency === ETH ? payAmount : fromWad(converter.usdToWei(payAmount))
 
+  const tierIdsFromIndices = (indices: number[]) => {
+    if (!rewardTiers) return []
+    return tierIdsFromTierIndices({
+      rewardTiers,
+      indices,
+    })
+  }
+
+  const sumSelectedTiers = (newSelectedIndices: number[]) => {
+    return newSelectedIndices.reduce((subSum, index) => {
+      return subSum + (rewardTiers?.[index]?.contributionFloor ?? 0)
+    }, 0)
+  }
```
**suggestion** if these dont rely on any internal state, we can move them outside of the component's scope.

tomquirk replied:
ah nevermind, they do.

tomquirk commented on src/components/NftRewards/NftRewardsSection.tsx:
```diff
@@ -75,40 +78,69 @@ export function NftRewardsSection() {
+      })
     },
-    [setPayMetadata, rewardTiers],
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
**question** why the need for this disable?

tomquirk commented on src/components/NftRewards/NftRewardsSection.tsx:
```diff
@@ -75,40 +78,69 @@ export function NftRewardsSection() {
+      setPayAmount?.(newPayAmount)
+      validatePayAmount?.(newPayAmount)
+    },
     // eslint-disable-next-line react-hooks/exhaustive-deps
```
**question** why the need for this disable?

tomquirk commented on src/components/NftRewards/NftRewardsSection.tsx:
```diff
@@ -75,40 +78,69 @@ export function NftRewardsSection() {
+  const deselectTier = useCallback(
+    tierIndex => {
+      const selectedIndex = selectedIndices.indexOf(tierIndex)
+      const newSelectedIndices =
```
**suggestion** perhaps a `filter` would be a simpler solution here.

```js
const newSelectedIndices = [...selectedIndices].filter(idx => idx !== selectedIndex)
```

tomquirk commented on src/components/NftRewards/NftRewardsSection.tsx:
```diff
@@ -117,14 +149,33 @@ export function NftRewardsSection() {
-      deselectTier()
     }
-  }, [payAmountETH, rewardTiers, deselectTier, selectTier, selectedIndex])
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
**question** why do we need this disable

tomquirk commented on src/components/NftRewards/NftRewardsSection.tsx:
```diff
@@ -117,14 +149,33 @@ export function NftRewardsSection() {
+  }, [selectedIndices, payAmountETH])
+
+  // sets pay input when selected nft's sum to greater than the current pay input amount
+  useEffect(() => {
```
**question** useEffect feels like an antipattern here. Is there a way we can execute this logic explictly then calling `setSelectedIndices` instead?

tomquirk commented on src/components/NftRewards/NftRewardsSection.tsx:
```diff
@@ -117,14 +149,33 @@ export function NftRewardsSection() {
+    ) {
+      setSelectedIndices([])
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
**question** why

tomquirk reviewed (changes requested):
Thanks for this @johnnyd-eth looking good, but left a few comments.

johnnyd-eth commented:
@tomquirk definitely a few unnecessary `// eslint-disable-next-line react-hooks/exhaustive-deps`s in there, which I've removed. 

But in the two instances it remains, including the `payAmountEth` or `selectedTierIds` causes fuckery. I added an extra comments clarifying the purposes of those `useEffect`s.

I tried to simplify this logic and combine into a single `useEffect`, but it felt a lot more intuitive and easy to break out into smaller `useEffect`s for particular events.
